koswat/strategies/order_strategy/order_strategy.py

```diff
@@ -91,28 +91,25 @@ def split_reinforcements() -> tuple[
             reverse=True,
         )
 
+        # Remove the reinforcements that are more expensive and less or equally restrictive than 1 of the others
         def check_reinforcement(
             pair: tuple[StrategyReinforcementInput, StrategyReinforcementInput],
         ) -> StrategyReinforcementInput | None:
-            # Only keep the less restrictive reinforcement if it is cheaper
             if (
-                pair[0].ground_level_surface > pair[1].ground_level_surface
-                and pair[0].base_costs < pair[1].base_costs
+                pair[0].base_costs > pair[1].base_costs
+                and pair[0].ground_level_surface >= pair[1].ground_level_surface
             ):
                 return pair[0]
             return None
 
-        # Check if the current (more expensive) reinforcement is more restrictive than the previous
-        # (the last needs to be appended as it is always kept)
-        _sorted_pairs = pairwise(_sorted + _last)
-        _kept = (
-            list(
-                filter(lambda x: x is not None, map(check_reinforcement, _sorted_pairs))
-            )
-            + _last
+        _pairs = product(_sorted, _sorted + _last)
+        _to_remove = set(
+            filter(lambda x: x is not None, map(check_reinforcement, _pairs))
         )
```

Comment:
computer wise I believe this to be better (and also simpler):

```python
        def check_reinforcement(
            pair: tuple[StrategyReinforcementInput, StrategyReinforcementInput],
        ) -> bool:
            return pair[0].base_costs > pair[1].base_costs and pair[0].ground_level_surface >= pair[1].ground_level_surface:

        for _pair in product(_sorted, _sorted + _last):
            if not check_reinforcement(_pair):
                continue
            _sorted.remove(_pair[0])

```

Reply:
The only downside of this solution is that a reinforcement could be flagged to be removed >1 time. `remove` will then give an exception, so in that case I need to add a check if it still exists.
With your solution the `__hash__` can be removed from `StrategyReinforcementInput`.
WDYT?

Reply:
I adjusted the solution in line with your suggestion. My comment on removing >1 time doesn't apply, as `_sorted` is adjusted during the loop.